add amber TI parser and improve the subsampling code

shuail opened the pull request:
Fixes #10 

1. add the amber TI parser to get the dhdl
2. Inside the statistical_inefficiency function under subsampling.py, switch to use the subsampleCorrelatedData in pymbar to generate the subsampled data points, the original code use the pandas slicing method which required to round the statistical inefficiency to an integer.

davidlmobley commented:
@shuail - this will need to come with tests. You'll probably want to put the data for the tests in `alchemtest` and then build in tests which rely on it here.

Note also that there is a "code coverage" analysis being run above which I presume is checking to see what portion of the code is being tested by the tests; since yours doesn't include tests, well, code coverage reduces dramatically and fails that test.

orbeckst commented:
As just discussed: 
1. add tests to alchemtest (open a PR there)
2. write tests (initially modelled on the Gromacs ones) that use the alchemtest data set

Ping me if _anything_ is unclear.

shuail commented:
@orbeckst just added the amber dataset in alchemtest and create a PR there. Added amber test here to use the files in alchemtest. Tested on my local computer to use the updated alchemtest, it runs fine. Finger crossed for travis once the alchemtest library get updated.

orbeckst commented:
I'll deal with PR  alchemistry/alchemtest#6 first to sort out the data side of things.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+import re
+import numpy as np
+import os
+
```
Use [logging](https://docs.python.org/3/library/logging.html) instead of `print`:
```
import logging

logger = logging.getLogger("alchemlyb.parsers.Amber")
```

See more comments below.

Note that in order to see the messages, a root logger has to be created elsewhere:
```python
import logging

def create_alchemlyb_logger(logfile="alchemlyb.log", name="alchemlyb"):
    """Create a logger that outputs to screen and to `logfile`"""

    logger = logging.getLogger(name)
    logger.setLevel(logging.DEBUG)

    # handler that writes to logfile
    logfile_handler = logging.FileHandler(logfile)
    logfile_formatter = logging.Formatter('%(asctime)s %(name)-12s %(levelname)-8s %(message)s')
    logfile_handler.setFormatter(logfile_formatter)
    logger.addHandler(logfile_handler)

    # define a Handler which writes INFO messages or higher to the sys.stderr
    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.INFO)
    # set a format which is simpler for console use
    formatter = logging.Formatter('%(name)-12s: %(levelname)-8s %(message)s')
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)

    return logger

rootlogger = create_alchemlyb_logger()
```
We might add this convenience function elsewhere in the library. I opened issue #34. 

Note that you only have to create the top level logger (here called `rootlogger`) once. Loggers are known globally in the interpreter, so you can "attach" to the root logger from anywhere else just by naming the new logger "alchemlyb.OTHER.STUFF".

shuail replied:
Yes, will switch to logging instead of printing. 

Just want to make sure that I understand the logic correctly, so what I need to do is to add 

```
import logging

logger = logging.getLogger("alchemlyb.parsers.Amber")
```

at the beginning of the code and start to logging like

logging.info ("Write some logging info here") is that correct?

orbeckst replied:
Yes, exactly.

(Just be aware that until a logger named "alchemlyb" has been created, no logging will be visible, but that's expected and desired behavior. With #34 I will add code to the library to start logging automatically.)

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        except IOError:
+            raise SystemExit('ERROR: cannot open file %s' % filename)
+        self.lineno = 0
+    def skip_lines(self, nlines):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+            if lineno > nlines:
+                return line
+        return None
+    def skip_after(self, pattern):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+            if match:
+                break
+        return self.fileh.tell() != self.filesize
+    def extract_section(self, start, end, fields, limit=None, extra='',
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+            else:                       # section may be incomplete
+                result.append(None)
+        return result
+    def __iter__(self):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        return result
+    def __iter__(self):
+        return self
+    def next(self):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+            raise StopIteration
+        # NOTE: can't mix next() with seek()
+        return self.fileh.readline()
+    def close(self):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+    def close(self):
+        """Close the filehandle."""
+        self.fileh.close()
+    def __enter__(self):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        self.fileh.close()
+    def __enter__(self):
+        return self
+    def __exit__(self, typ, value, traceback):
```
[PEP8](https://www.python.org/dev/peps/pep-0008/) formatting: leave a space between methods.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+class FEData(object):
+    """A simple struct container to collect data from individual files."""
+
+    #__slots__ = ['clambda', 't0', 'dt', 'T', 'gradients',
```
remove commented out code if not necessary

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+    with SectionParser(outfile) as secp:
+        line = secp.skip_lines(5) 
+        if not line:
+            print('  WARNING: file does not contain any useful data, '
```
turn `print` into 
```python
logger.warn("file %r does not contain any useful data, skipping file", outfile)
```
Add the filename to the error message; I assume that `outfile` was a filename.

orbeckst replied:
If you think that users should always see the message, use `warnings.warn(msg)` in addition to the logger.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+                      'ignoring file')
+            invalid = True
+        if not secp.skip_after('^   2.  CONTROL  DATA  FOR  THE  RUN'):
+            print('  WARNING: no CONTROL DATA found, ignoring file')
```
use logger

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        T, = secp.extract_section('temperature regulation:', '^$',
+                                 ['temp0'])
+        if not T:
+            raise SystemExit('ERROR: Non-constant temperature MD not '
```
Do not `SystemExit` in a library.

This is a library and calling code might want to handle it. Raise `NotImplementedError` instead.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        clambda, = secp.extract_section('^Free energy options:', '^$',
+                                        ['clambda'], '^---')
+        if clambda is None:
+            print('  WARNING: no free energy section found, ignoring file')
```
use logging

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+            invalid = True
+
+        if not secp.skip_after('^   3.  ATOMIC '):
+            print('  WARNING: no ATOMIC section found, ignoring file\n')
```
use logging

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+
+        t0, = secp.extract_section('^ begin time', '^$', ['coords'])
+        if not secp.skip_after('^   4.  RESULTS'):
+            print('  WARNING: no RESULTS section found, ignoring file\n')
```
use logging

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+        file_datum.t0 = t0
+        file_datum.dt = dt
+        file_datum.T = T
+        return file_datum
```
I would simplify to
```python
if invalid:
        return False

file_datum = FEData()
file_datum.clambda = clambda
file_datum.t0 = t0
file_datum.dt = dt
file_datum.T = T
return file_datum
```

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,265 @@
+
+def extract_dHdl(outfile, ):
+    file_datum = file_validation(outfile)
+    if file_validation(outfile):
```
rewrite the `if` to terminate early, the following would be clearer:
```python
if not file_validation(outfile):
   return None

finished = False
comps = []
...
return df
```

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 from pymbar.timeseries import statisticalInefficiency
 from pymbar.timeseries import detectEquilibration
+from pymbar.timeseries import subsampleCorrelatedData
```
I am not reviewing these changes here because they will be in a separate PR.

orbeckst commented on src/alchemlyb/tests/parsing/test_amber.py:
```diff
@@ -0,0 +1,21 @@
+from alchemtest.amber import load_simplesolvated
+
+
+def test_dHdl():
```
good start

shuail commented:
@orbeckst Thanks for all your comments, it's very helpful. I made three major changes 1, switch to logging in the amber parser and clean up the code (add docs, empty lines etc). 2, add test_ti_estimators_amber.py to test the ti_estimator for amber. 3, roll back the subsampling commit and will create a separated PR later.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,282 @@
+    df = convert_to_pandas(file_datum)        
+    return df
+
+#currently just check the code with a simple amber ti output file
```
remove the whole main section  (also, the `print` use python 2.7 syntax but we run under python 2 and python 3).

shuail replied:
Got it, the main section got removed now.

orbeckst commented:
This is starting to look good.

We would now like to get test coverage up. If you look at https://codecov.io/gh/alchemistry/alchemlyb/src/master/src/alchemlyb/parsing/amber.py (follow the codecov/project report and look for the overall coverage and navigate in the files) there are many lines in the function [parsing.amber.file_validation()](https://codecov.io/gh/alchemistry/alchemlyb/src/master/src/alchemlyb/parsing/amber.py#L171) that are not exercised by the tests. We would want to have tests that actually show that you catch these errors as expected. 

The best thing to do would be to include small files with errors and then just test the `file_validation()` function itself on these files. It is difficult to check that a log message was printed but you can test that `file_validation()` returns `False` for these cases. (As a sidenote, if you need to [test exceptions and warnings](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions) then pytest also makes this straight forward).

(This is where finalizing PRs becomes painful and boring but it is vital for the long-term survivability of the code that we check things *now* and are then assured that in the future everything will still work *or* we learn when it doesn't.)

Other comments on the code.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,277 @@
+    def __init__(self, filename):
+        """Opens a file according to its file type."""
+        self.filename = filename
+        with open(filename, 'r') as f:
```
This approach with `_MAGIC_CMPR` can be replace with `util.anyopen()`, see, for example https://github.com/alchemistry/alchemlyb/blob/master/src/alchemlyb/parsing/gmx.py#L143 . That's cleaner and more expandable. (I hadn't picked up on it in my earlier code reviews.)

(You can do it yourself or I am also happy to finish the PR as is and do this later and separately.)

shuail replied:
Got it, will try to switch to anyopen

shuail commented:
@orbeckst Got it, will add some testing for the file validation function. For the tested output files, I planned to created an invalidfiles folder under alchemtest/amber and use the alchemtest frame work in the alchemlyb. Does this sound good?

orbeckst commented:
Yes, good idea. Just check that the files are properly included when you run `pip install`, i.e., proper inclusions in `setup.py` (and perhaps `MANIFEST.in`... if we use it).

shuail commented:
@orbeckst Just finished adding the test code here and test files (alchemtest) to test the file validation function in amber parser and switched the file opener to anyopen. Hope the test will pass this time :)

orbeckst commented:
Looking good, just waiting for PR alchemistry/alchemtest#11.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+    return file_datum
+
+def extract_dHdl(outfile):
+    """Return gradients `dH/dl` from Amebr TI outputfile
```
typo: Amber

shuail replied:
Fixed

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+        self.filename = filename
+        try:
+            self.fileh = anyopen(self.filename, 'r')
+            self.filesize = os.stat(self.filename).st_size
```
This cannot be used for compressed files (see other comments).

I think this line can be deleted because it is useless.

orbeckst replied:
Well, thinking about this again: try it out, maybe it works and I am wrong.

shuail replied:
Yes, I think you are right, the filesize is not working here for zipped files, remove the filesize related detection in the code.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+            match = re.search(pattern, line)
+            if match:
+                break
+        return self.fileh.tell() != self.filesize
```
_Well, thinking about this again: try it out, maybe it works and I am wrong._

shuail replied:
Yes, I checked that the `self.filesize < self.fileh.tell()` for the zipped file and it will not be equal at the end of the file so I changed the `skip_after()` function to not use this detection but explicitly return `True` if the pattern was found along iterating the whole file

shuail commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+import re
+import pandas as pd
+import numpy as np
+import logging 
```
Got it, push a fixed

shuail commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+DVDL_COMPS = ['BOND', 'ANGLE', 'DIHED', '1-4 NB', '1-4 EEL', 'VDWAALS',
+              'EELEC', 'RESTRAINT']
+_FP_RE = r'[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?'
+_MAGIC_CMPR = {
```
removed

shuail commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+        yield first
+
+    while it:
+        yield it.next()
```
Yes, change it to next(it)

orbeckst commented:
I merged alchemistry/alchemtest#11 so you should now have the appropriate test files available for Travis CI. I will manually restart the tests. You don't need to do anything until the test results come back.

shuail commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,270 @@
+        """Read next line of the filehandle and check for EOF."""
+        self.lineno += 1
+        curr_pos = self.fileh.tell()
+        if curr_pos == self.filesize:
```
Yes, that's a better solution. Sorry I didn't think that part much as migrating the code from alchemical-analysis. It looks like that this test successfully catch the untested part of the original code!

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,264 @@
+import pandas as pd
+import numpy as np
+
+from util import anyopen
```
Needs to be 
```python
from .util import anyopen
```

orbeckst replied:
https://travis-ci.org/alchemistry/alchemlyb/jobs/301541570

shuail replied:
Oops, yes, fixed this

shuail commented:
@orbeckst thanks for your comments! Yes, after I switched to the zipped files, the end of file(EOF) detection in the `skip_after` function will not work, so I changed that part and also the next() function to make it suitable for the zipped file. The `anyopen` actually works fine here.

orbeckst commented:
One small change to get it to pass under python 3.

orbeckst commented:
Once the tests pass, we have the coverage. Assuming that this will look good, we need to make sure that it can be merged and I want to clean up the PR, i.e., merging related commits and removing the subsampling code that was added and deleted. I can do this locally with `git rebase`. 

A consequence for you will be that you also have to rebase your fork because you have been working from master instead of a branch. I can tell you what I would do if you want to know.

orbeckst commented:
This looking very good. The only way that you can sensibly improve the coverage is by adding a test for the function
```python
parsing.amber.any_none(sequence)
```
Basically, make sure that the [line 42](https://codecov.io/gh/alchemistry/alchemlyb/compare/8915ece9daa2ad17094b2e61565bcfe128a374c7...49cc45cf3f310e1a6eff1d4c15f6fa2c17adbd09/diff#D1-42) is hit. Any of the other remaining checks are fairly standard defensive programming and I am not even sure how to test them properly.

orbeckst commented on src/alchemlyb/parsing/amber.py:
```diff
@@ -0,0 +1,264 @@
+
+    for element in sequence:
+        if element is None:
+            return True
```
This looking very good. The only way that you can sensibly improve the coverage is by adding a test for the function
```python
parsing.amber.any_none(sequence)
```
Basically, make sure that the [line 42](https://codecov.io/gh/alchemistry/alchemlyb/compare/8915ece9daa2ad17094b2e61565bcfe128a374c7...49cc45cf3f310e1a6eff1d4c15f6fa2c17adbd09/diff#D1-42) is hit. Any of the other remaining checks are fairly standard defensive programming and I am not even sure how to test them properly.

shuail replied:
Got a test function added for this line

orbeckst commented:
@shuail looking good, I'll do the rebase... hopefully done before I have to run to class...

orbeckst commented:
I created PR #36 to finalize.
